src/machines/java/com/enderio/machines/common/blockentity/task/CraftingMachineTask.java

```diff
@@ -0,0 +1,228 @@
+package com.enderio.machines.common.blockentity.task;
+
+import com.enderio.core.common.recipes.OutputStack;
+import com.enderio.machines.common.io.item.MachineInventory;
+import com.enderio.machines.common.io.item.MultiSlotAccess;
+import com.enderio.machines.common.io.item.SingleSlotAccess;
+import com.enderio.machines.common.recipe.MachineRecipe;
+import net.minecraft.nbt.CompoundTag;
+import net.minecraft.nbt.ListTag;
+import net.minecraft.nbt.Tag;
+import net.minecraft.resources.ResourceLocation;
+import net.minecraft.world.Container;
+import net.minecraft.world.item.ItemStack;
+import net.minecraft.world.level.Level;
+import org.jetbrains.annotations.NotNull;
+import org.jetbrains.annotations.Nullable;
+
+import java.util.ArrayList;
+import java.util.List;
+
+// TODO: A recipe interface that doesn't require power :)
+public abstract class CraftingMachineTask<R extends MachineRecipe<C>, C extends Container>
+    implements IMachineTask {
+
+    protected final Level level;
+    protected final MachineInventory inventory;
+    protected final C container;
+    protected final MultiSlotAccess outputSlots;
+
+    @Nullable
+    private R recipe;
+
+    private int progressMade;
+    private int progressRequired;
+
+    private boolean hasConsumedInputs;
+    private boolean hasDeterminedOutputs;
+
+    private List<OutputStack> outputs = List.of();
+
+    private boolean isComplete;
+
+    public CraftingMachineTask(@NotNull Level level, MachineInventory inventory, C container, MultiSlotAccess outputSlots, @Nullable R recipe) {
+        this.level = level;
+        this.inventory = inventory;
+        this.container = container;
+        this.outputSlots = outputSlots;
+        this.recipe = recipe;
+    }
+
+    public MachineInventory getInventory() {
+        return inventory;
+    }
+
+    @Nullable
+    public R getRecipe() {
+        return recipe;
+    }
+
+    // region Abstract Implementation
+
+    protected abstract void consumeInputs(R recipe);
+
+    protected abstract int makeProgress(int remainingProgress);
```

Comment:
The rationale behind "progress" is to allow machines that may not necessarily consume power to still take advantage of the crafting tasks. 90% of time, "progress" will just be "energy".